pkg/reconciler/broker/broker.go

```diff
@@ -0,0 +1,505 @@
+	projectID string
+}
+
+// Check that Reconciler implements Interface
+var _ brokerreconciler.Interface = (*Reconciler)(nil)
+var _ brokerreconciler.Finalizer = (*Reconciler)(nil)
+
+var brokerGVK = brokerv1beta1.SchemeGroupVersion.WithKind("Broker")
+
+func (r *Reconciler) ReconcileKind(ctx context.Context, b *brokerv1beta1.Broker) pkgreconciler.Event {
+	if err := r.reconcileBroker(ctx, b); err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling broker", zap.Error(err))
+		return fmt.Errorf("failed to reconcile broker: %w", err)
+		//TODO instead of returning on error, update the data plane configmap with
+		// whatever info is available. or put this in a defer?
+	}
+
+	if err := r.reconcileTriggers(ctx, b); err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling triggers", zap.Error(err))
+		return fmt.Errorf("failed to reconcile triggers: %w", err)
+	}
+
+	logging.FromContext(ctx).Info("targetsConfig", zap.Any("cfg", r.targetsConfig.String()))
+
+	return pkgreconciler.NewEvent(corev1.EventTypeNormal, brokerReconciled, "Broker reconciled: \"%s/%s\"", b.Namespace, b.Name)
+}
+
+func (r *Reconciler) FinalizeKind(ctx context.Context, b *brokerv1beta1.Broker) pkgreconciler.Event {
+	logger := logging.FromContext(ctx)
+	logger.Debug("Finalizing Broker", zap.Any("broker", b))
+
+	// Reconcile triggers so they update their status
+	if err := r.reconcileTriggers(ctx, b); err != nil {
+		logger.Error("Problem reconciling triggers", zap.Error(err))
+		return fmt.Errorf("failed to reconcile triggers: %w", err)
+	}
+
+	if err := r.deleteDecouplingTopicAndSubscription(ctx, b); err != nil {
+		return fmt.Errorf("Failed to delete Pub/Sub topic: %w", err)
+	}
+
+	// TODO should we also delete the broker from the config? Maybe better
+	// to keep the targets in place if there are still triggers. But we can
+	// update the status to UNKNOWN and remove address etc. Maybe need a new
+	// status DELETED or a deleted timestamp that can be used to clean up later.
+
+	return pkgreconciler.NewEvent(corev1.EventTypeNormal, brokerFinalized, "Broker finalized: \"%s/%s\"", b.Namespace, b.Name)
+
+}
+
+func (r *Reconciler) reconcileBroker(ctx context.Context, b *brokerv1beta1.Broker) error {
+	logger := logging.FromContext(ctx)
+	logger.Debug("Reconciling Broker", zap.Any("broker", b))
+	b.Status.InitializeConditions()
+	b.Status.ObservedGeneration = b.Generation
+
+	// Create decoupling topic and pullsub for this broker. Ingress will push
+	// to this topic and fanout will pull from the pull sub.
+	if err := r.reconcileDecouplingTopicAndSubscription(ctx, b); err != nil {
+		return fmt.Errorf("Decoupling topic reconcile failed: %w", err)
+	}
+
+	//TODO in a webhook annotate the broker object with its ingress details
+	// Route to shared ingress with namespace/name in the path as the broker
+	// identifier.
+	b.Status.SetAddress(&apis.URL{
+		Scheme: "http",
+		Host:   names.ServiceHostName("broker-ingress", system.Namespace()),
```

Comment:
Should `broker-ingress` be a const? I assume it's also used when creating the service.